lib/grandpa/catch-up.go

```diff
@@ -0,0 +1,289 @@
+// Copyright 2021 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"errors"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/crypto/ed25519"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+const catchUpResponseTimeout = time.Second * 5
+
+type catchUp struct {
+	lock sync.Mutex // applied on requestsSent and bestResponse
```

Comment:
I would suggest to use RWMutex to maps, use `lock.RLock()` while reading and `lock.Lock()` while writing to the map
```suggestion
	lock sync.RWMutex // applied on requestsSent and bestResponse
```